src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2356,57 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static void ucp_worker_keepalive_deactivate(ucp_worker_h worker)
+{
+    uct_worker_progress_unregister_safe(worker->uct,
+                                        &worker->keepalive.cb_id);
+}
+
+static unsigned ucp_worker_keepalive_cb(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    double now          = ucs_get_accurate_time();
+    unsigned cnt;
+    ucp_ep_ext_gen_t *ep_ext;
+    ucp_ep_h ep;
+    ucs_list_link_t *iter_begin;
+
+    if ((now - worker->keepalive.last_shot) <
+        worker->context->config.ext.keepalive_timeout) {
+        return 0;
+    }
+
+    worker->keepalive.last_shot = now;
+    cnt                         = 0;
+    iter_begin                  = worker->keepalive.iter;
```

Comment:
what if ep is removed? it can be invalid

Reply:
when EP is removed - iterator is shifted to begin of list

Reply:
why not to next ep?

Reply:
because ep is removed from list prior to update iterator

Reply:
i see the problem but this solution can don't start keepalive for tail of the eps list if at least 1 EP is actively created/deleted, right?

Reply:
iterator is moved only in case if it was pointed to removed EP